File: 021_merge_two_sorted.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if (list1 is None):
            return list2
        if (list2 is None):
            return list1

        el = None
        if (list1.val > list2.val):
            el = list2.val
            list2 = list2.next
        else:
            el = list1.val
            list1 = list1.next

        head = ListNode(el)
        curr = head
        while (list1 != None or list2 != None):
            if (list1 is None):
                curr.next = ListNode(list2.val)
                list2 = list2.next
            elif (list2 is None):
                curr.next = ListNode(list1.val)
                list1 = list1.next
            else:
                el = None
                if (list1.val < list2.val):
                    el = list1.val
                    list1 = list1.next
                else:
                    el = list2.val
                    list2 = list2.next
                curr.next = ListNode(el)
            curr = curr.next
        return head

File: 021_merge_two_sorted.ts
function mergeTwoLists(
    list1: ListNode | null,
    list2: ListNode | null
): ListNode | null {
    const dummy1 = new ListNode(0, list1);
    let prev: ListNode | null = dummy1;

    while (list1 !== null && list2 !== null) {
        if (list1.val <= list2.val) {
            list1 = list1.next;
            prev = prev!.next;
        } else {
            let temp = list2.next;
            prev!.next = list2;
            list2.next = list1;
            list2 = temp;
            list1 = prev!.next;
        }
    }

    while (list2) {
        prev!.next = list2;
        prev = prev!.next;
        list2 = list2.next;
    }

    return dummy1.next;
}
